### ISSUE TITLE
register_post_type() $args standards?

### ISSUE THREAD

**Issue by AlchemyUnited** [justintadlock/content-type-standards#32]
Perhaps this is the wrong place and/or it's too late (as this idea - i.e., content type standard- while great and ambitious - looks to be dormant?).

In any case, backing up a bit, what about a set of standards / defaults for the $args of a CPT? 

I mean, instead of having to define the same settings over and over - say nothing of just remembering what's what :) - wouldn't it make sense to have list (e.g., array) where you can say "I want key X" and just plug that in?

Mind you this doesn't include the $labels. Obviously, that's gonna vary. And since it's an array there would still be opportunity for customization. I'm talking about, for example, a $key = 'slider' (ok, sorry, slider, bad example, lol) where the $args are going to be pretty much standard. 

From a pure n00bs POV wouldn't it make a lot of sense to be able to dial up a CPT using some known set of $args, plug that in, and then just doing the labels? Each standard could have brief description of what it might be good for, or not good for. 

Just a thought. 

p.s. I've been wanting to do this myself but I figured this might be a good spot to ask for some help? And get the idea off the ground?

**Comment by AlchemyUnited** [comment]
I might as well add this:

	'post_type'   => false,
	'text_domain' => 'TODO',
	'singular'    => 'TODO singular',
	'plural'      => 'TODO plural',
	'description' => 'TODO description',
	'rpt_labels'  => array(),
	'rpt_args'    => array(),
	'rpt_args_menu_icon' = 'dashicons-thumbs-down',
	'taxonomies'  => array(),

I've hacked out a class for rpt() and ultimately I've been able to abstact it down to the above list. That is, fill in those blanks and the rest just happens :)

Note: rpt_labels is just to override the's class' more generic approach that plugins in singular and plural on the fly. 

The big MIA is the rpt_args. Those can and do vary depending on the CPT. That said, my sense is, 80% of all CPTs probably come down to a handful of different sets of $args. Having those standardized - imho - would be a helpful nudge forward. 

Anyone?

**Comment by justintadlock** [comment]
Core WP already sets default `$args`.  So, anything that you don't want to change can simply be left out.  That's your _standard_ for that.

If you just need a starter template for CPTs, here's one: https://gist.github.com/justintadlock/6552000

Not sure what we need beyond that.  CPT registration is just a very small part of things.  I wouldn't worry about trying to build upon what's already in core as far as registration goes.  CPT implementation is the important stuff.  Plus, it's where all the cool things happen.

**Comment by AlchemyUnited** [comment]
Hi @justintadlock - Happy New Year. 

Yes. I'm aware of the core defaults :)

What I'm suggesting is that there are a finite number (read: 80%) of CPT type. I mean, if want a CPT for Personnel that will be on the front-end all those settings would be the same across 80% of all implementations. 

On the other hand, there's might be an Admin only CPT. 

I agree. Minor. But it's not standardized. 

For example, take a some Content Type Standard X and apply it to some CPT Type Y + a few specific args (as above) and...wait for it...DONE. 

Yeah. CPT Types are just one (small) part. But if the arc is here is about KISS + DRY + Standards then why not just get at least the CPT Types done?

**Comment by AlchemyUnited** [comment]
p.s. That gist is helpful but I'm suggesting it be taken a step further. For example, this is how I'm doing the labels:

			$singular = esc_attr($this->_singular);
			$plural   = esc_attr($this->_plural);

			$arr_labels = array(
				'name'                  => _x( $plural, 'Post Type General Name', $this->_text_domain ),
				'singular_name'         => _x( $singular, 'Post Type Singular Name', $this->_text_domain ),
				'menu_name'             => __( $plural, $this->_text_domain ),
				'name_admin_bar'        => __( $plural, $this->_text_domain ),
				'archives'              => __( $singular, $this->_text_domain ) . ' ' .  __('Archives', $this->_text_domain) ,
				'attributes'            => __( $singular, $this->_text_domain ) . ' ' . __('Attributes', $this->_text_domain),
				'parent_item_colon'     => __( $singular, $this->_text_domain ) . ':',
				'all_items'             => __( 'All', $this->_text_domain ) . ' ' . $plural,
				'add_new_item'          => __('Add New', $this->_text_domain) . ' ' .  __( $singular, $this->_text_domain ),
				'add_new'               => __( 'Add New', $this->_text_domain ),
				'new_item'              => __( 'New', $this->_text_domain ) . ' ' . $singular,
				'edit_item'             => __( 'Edit', $this->_text_domain ) . ' ' . $singular,
				'update_item'           => __( 'Update', $this->_text_domain ) . ' ' . $singular,
				'view_item'             => __( 'View', $this->_text_domain ) . ' ' . $singular,
				'view_items'            => __( 'View', $this->_text_domain ) . ' ' . $plural,
				'search_items'          => __( 'Search', $this->_text_domain ) . ' ' . $plural,
				'not_found'             => __( 'Sorry. Search found no', $this->_text_domain ) . ' ' . $plural,
				'not_found_in_trash'    => __( 'Not found in Trash', $this->_text_domain ),
				'featured_image'        => __( 'Featured Image', $this->_text_domain ),
				'set_featured_image'    => __( 'Set featured image', $this->_text_domain ),
				'remove_featured_image' => __( 'Remove featured image', $this->_text_domain ),
				'use_featured_image'    => __( 'Use as featured image', $this->_text_domain ),
				'insert_into_item'      => __( 'Insert into this', $this->_text_domain ) . ' ' . $singular,
				'uploaded_to_this_item' => __( 'Uploaded to this', $this->_text_domain ) . ' ' . $singular,
				'items_list'            => $plural . ' ' . __( 'list', $this->_text_domain ),
				'items_list_navigation' => $plural . ' ' . __( 'list navigation', $this->_text_domain ),
				'filter_items_list'     => $plural . __( ': Filter list', $this->_text_domain ),
			);

In short, once I know singular and plural the labels are done. No need to even muck with that (unless I feel that that default isn't the right fit. Then yes, I can override. But at the end of the day the $label can be done by just knowing singular and plural. 

Minor? Yes, I agree again. But it's also KISS + DRY. And one less minor / stupid thing to worry about. One less minor / stupid thing to test, etc. 

I'm just wanting to dumb all this down to the simplist of simple. I don't want to code. I simply want to configure. Make sense?

**Comment by AlchemyUnited** [comment]
I'm thinking of it in terms of "objects." Get a CPT "object" and slap a Content Type "object" on top. Done! In that context I think it's possible to distill (i.e., abstract) down to a minimal number of args per CPT+CT. Again, less actually coding and more simple configuring (because classes handle most of the dirty work).

**Comment by AlchemyUnited** [comment]
@justintadlock  - two things:

1) I think the difference in POV is you're thinking "commercial" plugins and I'm thinking (one-off-y) custom work. For example, you have the gist with a bunch of functions. I have a class that does all my heavy lifting. 

I'm not saying one is right / wrong, just that it's two (slightly) different playing fields. That said, I can easily stuff my rpt class into anything "commericial" by just changing the namespace. Not ideal, but it's doable. 

2) On that note - simply as an FYI - this is what my $args look like for my rpt() class:

	$arr_args        = array(
		'active'             => true,
		'active_filters'     => false,
		'post_type'          => 'faq',
		'text_domain'        => 'todo',
		'singular'           => 'FAQ',
		'plural'             => 'FAQs',
		'description'        => 'Frequently Asked Questions',
		'rpt_labels'         => array(),
		'rpt_args'           => array(
			'show_ui' => true,
		),
		'args_menu_position' => 20,
		'args_menu_icon'     => 'dashicons-phone',
		'args_supports'      => array( 'title', 'editor' ),
		'args_taxonomies'    => array(),
	);

That's it. All the other "code" is buried in the class. DRY :)

Yes, the last 4 could be part of rpt_args() but I presume they are elements of what defines a "minimal viable post type" :) 

Essentially, this is the only "code" I have to deal with to fire up a CPT. Fill on these "blanks" and you have a CPT. Now all I need is CMB2 to define the actual fields - which I'll also have a library of (in the spirit of this standards project). To "add on" the "what type of CPT is this?" would make things even easier. 

CPT $args + CPT type + CMB2 = BOOM! 

I hope this helps. Thanks again.

**Comment by justintadlock** [comment]
I'm going to close this ticket because I don't think it's right for this project.  I'll try to explain some of this below.

This project deals directly with plugins released to the public, so there's no different points of view.  Custom projects can certainly use these standards.  However, they are outside the scope of this project.

Anyway, on to your code examples.  Things like this are broken:

```
'menu_name' => __( $plural, $this->_text_domain ),
```

First, you can't translate `$plural` because it's a variable.  Variables should never be inside of a gettext function.

Second, your textdomain must be a hardcoded string.  Otherwise, many translation tools cannot properly parse the strings.  It's important to note that most tools are not running PHP.  So, `$this->_text_domain` doesn't mean anything to those tools.

I highly recommend reading this post: https://markjaquith.wordpress.com/2011/10/06/translating-wordpress-plugins-and-themes-dont-get-clever/

> I'm just wanting to dumb all this down to the simplist of simple. I don't want to code. I simply want to configure. Make sense?

Yep, that makes sense.  That's why WordPress has a registration function.  It's for the configuration of your CPT.

Outside of registration, if you're truly using CPTs to their full potential, you'll be doing lots of coding.  It's the implementation that matters.

Trying to put a configuration wrapper (`rpt()`) around a configuration wrapper (`register_post_type()`) might make sense for some of your custom projects if you have a lot of CPTs with similar configuration, but it's not something I'd really recommend for most developers.  It's just an extra layer that they'd have to learn and deal with.

> I'm thinking of it in terms of "objects." Get a CPT "object" and slap a Content Type "object" on top. Done! In that context I think it's possible to distill (i.e., abstract) down to a minimal number of args per CPT+CT. Again, less actually coding and more simple configuring (because classes handle most of the dirty work).

That's the point where I'd vehemently disagree.  Again, this might be the case for your own, custom projects.  However, in my experience, the configuration is so varied that it doesn't make sense to add that second layer in an attempt to cut back on a few lines of code in my main registration function.

**Comment by AlchemyUnited** [comment]
Mind if I reply? :)

> This project deals directly with plugins released to the public, so there's no different points of view. Custom projects can certainly use these standards. However, they are outside the scope of this project.

100% understood. But it's walking a fine line. Read: Why should there be a difference? Especially at the level you seem to be approaching this on?

> First, you can't translate $plural because it's a variable. Variables should never be inside of a gettext function.

Oh. Opps. Good to know. Thanks. Shame really. 

>  That's why WordPress has a registration function. 

> It's just an extra layer that they'd have to learn and deal with.

Actually. I see it differently. Sure CPTs are obvious to you and anyone with loads of experience. If you're not in that camp then its overwhelming, TMI, etc.  To abstract out the essentials, and let a black box (i.e., class) do the rest...well, why doesn't that make sense? KISS is bad? DRY is bad? Ya lost me brother :(

>  the configuration is so varied 

I bet you $100 that if you looked back you'd see patterns; patterns that I bet cover 80% of use cases. Patterns that if you had a UI you could just drag a "CPT type" (object) onto the main object and you'd be done. (Note: For a n00b that would be MASSIVE.). Add in the "which standard" object and...Done! 

Even if it wasn't D&D it could be a couple selects - a la Generate WP - and then spit out a copy & paste block of code. 

```
<sidebar>

Pardon me for saying this but if you want someone to use something you have to make it as ez as possible. If you want people to use these content  type standards then there's has to be a plus, an advantage FOR THEM. 

I agree with you. It would be good for the "community." But that is a hard sell to devs who are underpaid, don't have the time, etc. They will be asking - as everyone does - **"What's in it for me?"**

In short, imho, it's a package deal. You see otherwise. I think history is on my side :)

</sidebar>
```

I'm not disputing fringe cases. Or an 80%'er that might need one or two properties tweaked. But to say implementation is constant wheel code reinvention when it's a handful of objects makes me uncomfortable (in 2017).

You're coding when you could be configuring. You're coding when someone junior could be configuring. You're coding needs to be tested. You're hand coding stuff that could (read: should) be assembled or at least just generated. 

I wish I had that kinda time to waste :)

I also wish everyone had the depth and breadth of experience you do. (Hint: They do not.)

Thanks for listening. Once you get around to thinking about adoption we can restart this conversation :)

**Comment by justintadlock** [comment]
> 100% understood. But it's walking a fine line. Read: Why should there be a difference? Especially at the level you seem to be approaching this on?

There doesn't have to be a difference.  It's simply what this project is about.  It's about if I build Portfolio Plugin A and Joe builds Portfolio Plugin B, the user would still have their data when they switched plugins.  What we're doing is naming things (e.g., projects in a portfolio should be named `portfolio_project`).  That's what it all boils down to.

This doesn't really matter for custom work.  Of course, custom work can and probably should follow any standards.  However, custom work doesn't have the issues that this project is attempting to solve.

Anything outside of a standardized naming scheme is really outside the scope of the project.  How things are configured doesn't really matter so much as long as you have the naming schemes right.

> Pardon me for saying this but if you want someone to use something you have to make it as ez as possible. If you want people to use these content  type standards then there's has to be a plus, an advantage FOR THEM. 

I'm in no way trying to sell the project's value.  I thought the idea had some merit but was not that necessary.  People asked for it, so I created a centralized place for people to create some naming standards.

As far as ease of use goes, it's copy and paste.

If I go build some entirely new system on top of the existing, known system, it creates a learning curve for both beginners and experienced devs.  If we use the universal system (i.e., WordPress functions), it keeps everyone on the same page.  There's existing documentation around the Web as well.

There's certainly a place for a wrapper script like you're describing.  There's probably a few already out there. I'd even be happy to look over your code if you wanted to start a public project like that. This is just not the place for it.

**Comment by AlchemyUnited** [comment]
@justintadlock  

p.s. I did some poking around and maybe you can clarify....

> re: 'menu_name' => __( $plural, $this->_text_domain ),
> 
> "First, you can't translate $plural because it's a variable. Variables should never be inside of a gettext function."

But this isn't the gettext() function, __() is an alias, yes? By the time it gets to gettext() it is for all intents and purposes a string, yes? As the __() function just passes its args on to gettext() Same with the text_domain. 

Futhermore __() doesn't even deal directly with gettext(), see:

```
	/**
79	 * Retrieve the translation of $text.
80	 *
81	 * If there is no translation, or the text domain isn't loaded, the original text is returned.
82	 *
83	 * *Note:* Don't use translate() directly, use __() or related functions.
84	 *
85	 * @since 2.2.0
86	 *
87	 * @param string $text   Text to translate.
88	 * @param string $domain Optional. Text domain. Unique identifier for retrieving translated strings.
89	 *                       Default 'default'.
90	 * @return string Translated text
91	 */
92	function translate( $text, $domain = 'default' ) {
93	        $translations = get_translations_for_domain( $domain );
94	        $translations = $translations->translate( $text );
95	
96	        /**
97	         * Filters text with its translation.
98	         *
99	         * @since 2.0.11
100	         *
101	         * @param string $translations Translated text.
102	         * @param string $text         Text to translate.
103	         * @param string $domain       Text domain. Unique identifier for retrieving translated strings.
104	         */
105	        return apply_filters( 'gettext', $translations, $text, $domain );
106	}

```
Otherwise - and maybe I'm mis-thinking my understanding of WP / OOP - there's no way to internationalize such things within a class. That doesn't make sense to me. Of course, that might be the way it is :) But in terms of (WP / OOP) dev, it feels pretty awkward / odd to me. 

Regardless, it's not a deal breaker for what I envision. I'm just posting this for completeness. And ideally - if I'm wrong - to be corrected. That last thing I need in my head is more mis-information. 

BTW, that code doesn't bomb. Of course, that doesn't mean what I did works, but in terms of not blowing up, it does not blow up. That's kinda unexpected results considering it's, per you, broken. 

Thanks again brother. Keep the faith!

**Comment by justintadlock** [comment]
No, it's not THE `gettext()` function.  It's A gettext function.  It's in the family of WP's gettext functions.  It's just what they're called.  I could've also said an internationalization function.

But, no, it won't blow up the site.  It'll function.  It just won't work for the people who need to translate (can't translate variable).
